File: 0042-trapping-rain-water/0042-trapping-rain-water.swift
class Solution {
    func trap(_ height: [Int]) -> Int {
        var left = 0, leftMax = 0, rightMax = 0, totalWaterCollected = 0
        var right = height.count - 1

        while left<right{
            if height[left] < height[right]{
                if height[left] >= leftMax{
                    leftMax = height[left]
                }
                else{
                    totalWaterCollected += leftMax - height[left]
                }
                left+=1
            }
            else{
                if height[right]>=rightMax{
                    rightMax = height[right]
                }
                else{
                    totalWaterCollected += rightMax - height[right]
                }
                right -= 1
            }
        }

        return totalWaterCollected
    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.java
//two-pointer approach
//time complexity: O(n), space complexity: O(1)
class Solution {
    public int trap(int[] height) {
        int left = 0, right = height.length - 1;
        int maxLeft = 0, maxRight = 0;
        int totalTrapped = 0;

        while (left < right){
            if(height[left] < height[right]){ // process left side
                if(height[left] >= maxLeft){
                    maxLeft = height[left];
                }
                else{
                    totalTrapped += maxLeft - height[left];
                }
                left++;
            }
            else{ //height[left] > height[right] // process right side
                if(height[right] >= maxRight){
                    maxRight = height[right];
                }
                else{
                    totalTrapped += maxRight -height[right];
                }
                right--;
            }
        }
        return totalTrapped;
    }
}
